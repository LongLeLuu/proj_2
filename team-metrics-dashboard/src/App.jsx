import "./output.css"; // Import global styles
import { useEffect, useState } from "react";
import Header from "./components/Header.jsx";
import SearchForm from "./components/SearchForm.jsx";
import Footer from "./components/Footer.jsx";
import { Outlet } from "react-router-dom";
import { Link } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { MousePointer2 } from "lucide-react";

/*UI components*/
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";
import { Skeleton } from "@/components/ui/skeleton";

function App() {
  const navigate = useNavigate();
  const [responseData, setResponseData] = useState(null);
  const [theme, setTheme] = useState("dark"); // Default to dark mode
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  //Saving preferred viewing mode currently does not work..
  useEffect(() => {
    // Check for saved theme in localStorage
    const savedTheme = localStorage.getItem("theme");
    console.log(savedTheme);
    if (savedTheme) {
      setTheme(savedTheme);
    }
  }, []);

  //Once data is fetched, redirect to first array of data
  //(and disable loading animation)
  useEffect(() => {
    if (responseData) {
      navigate("/0");
    }
  }, [responseData]);

  const handleLinkChange = (e) => {
    //handles option changes, modifies URL based on selected option
    if (e) {
      navigate(e); //triggers re-render of Data component as it's dependent on dynamic parameter of :index
    }
  };

  /*The theme can be manipulated using state.. 
  but effects are necessary to directly change the DOM 
  depending on dark/light mode
  
  Changes in state will rerender the whole app, while 
  changes in effect are dependent upon the parameters in the 
  dep. array
  */
  useEffect(() => {
    // Apply theme class to root element
    document.documentElement.classList.remove(
      theme === "dark" ? "light" : "dark"
    );
    document.documentElement.classList.add(theme);

    // Save theme to localStorage
    localStorage.setItem("theme", theme);
  }, [theme]);

  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === "dark" ? "light" : "dark"));
  };

  return (
    <div className="App box-border">
      <Header
        toggleTheme={toggleTheme}
        currentTheme={theme}
        className="header"
      />
      <div className="container">
        <main className={`main-content w-screen`}>
          <SearchForm
            setResponseData={setResponseData}
            loading={loading}
            setLoading={setLoading}
            setError={setError}
          />
          {loading && (
            <div className="flex flex-wrap gap-16 items-center justify-center">
              <div className="flex flex-col space-y-3">
                <Skeleton className="h-[250px] xl:w-[800px] lg:w-[450px] md:w-[500px] sm:w-[350px] rounded-xl" />
                <div className="space-y-2">
                  <Skeleton className="h-4 w-[350px]" />
                  <Skeleton className="h-4 w-[300px]" />
                </div>
              </div>
              <div className="flex flex-col space-y-3">
                <Skeleton className="h-[250px] xl:w-[800px] lg:w-[450px]  md:w-[500px] sm:w-[350px] rounded-xl" />
                <div className="space-y-2">
                  <Skeleton className="h-4 w-[350px]" />
                  <Skeleton className="h-4 w-[300px]" />
                </div>
              </div>
            </div>
          )}

          {responseData && !loading && (
            <>
              <div className="w-screen">
                <Select onValueChange={handleLinkChange}>
                  <SelectTrigger className="w-3/12 ml-4">
                    <SelectValue
                      placeholder={
                        <div className="flex items-center ">
                          Select Subject/Owner... <MousePointer2 height={15} />
                        </div>
                      }
                    />
                  </SelectTrigger>
                  <SelectContent>
                    {responseData.intersect &&
                      responseData.data.map((arr, index) => {
                        return (
                          <SelectItem key={index} value={`${index}`}>
                            Subject: {arr[0]["reason"]}
                          </SelectItem>
                        );
                      })}
                    {!responseData.intersect &&
                      responseData.data.map((arr, index) => {
                        if (index + 1 > responseData.subjectSplit) {
                          return (
                            <SelectItem key={index} value={`${index}`}>
                              Owner: {arr[0]["name"]}
                            </SelectItem>
                          );
                        } else {
                          return (
                            <SelectItem key={index} value={`${index}`}>
                              Subject: {arr[0]["reason"]}
                            </SelectItem>
                          );
                        }
                      })}
                  </SelectContent>
                </Select>
                {<Outlet context={[responseData.data]} />}
              </div>
            </>
          )}
          {error && <p>{error}.</p>}
        </main>
      </div>
      <Footer />
    </div>
  );
}

export default App;
