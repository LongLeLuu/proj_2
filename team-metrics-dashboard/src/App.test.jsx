// App.test.jsx

import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";

describe("App component", () => {
  it("Renders snapshot", () => {
    // since screen does not have the container property, we'll destructure render to obtain a container for this test
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("Renders Button", () => {
    render(<App />);
    const button = screen.getByRole("button", { name: /advanced settings/i });
    expect(button).toBeInTheDocument();
  });

  it("Advanced Settings toggles div properly", async () => {
    const user = userEvent.setup();

    render(<App />);
    const button = screen.getByRole("button", {
      name: /advanced settings/i,
    });

    await user.click(button);
    const startDate = screen.getByText(/start date:/i).innerHTML;
    const endDate = screen.getByText(/end date:/i).innerHTML;
    const intersect = screen.getByText(/intersect/i).innerHTML;

    expect(startDate).toMatch("Start Date");
    expect(endDate).toMatch("End Date:");
    expect(intersect).toMatch("Intersect");
  });
});
