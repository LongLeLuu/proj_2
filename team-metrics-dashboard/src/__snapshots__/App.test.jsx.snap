// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > Renders snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="header"
    >
      <h1>
        Team Metrics Dashboard
      </h1>
      <div
        class="images"
      >
        <img
          alt="Lightbulb"
          height="10"
          src="/src/assets/lightbulb.png"
          width="10"
        />
        <img
          alt="Moon"
          height="10"
          src="/src/assets/moon.png"
          width="10"
        />
      </div>
    </header>
    <div
      class="container"
    >
      <aside
        class="sidebar"
      >
        <nav>
          <ul>
            <li>
              <a
                href="#overview"
              >
                Overview
              </a>
            </li>
            <li>
              <a
                href="#commits"
              >
                Commits
              </a>
            </li>
            <li>
              <a
                href="#code-quality"
              >
                Code Quality
              </a>
            </li>
            <li>
              <a
                href="#reviews"
              >
                Reviews
              </a>
            </li>
          </ul>
        </nav>
      </aside>
      <main
        class="main-content"
      >
        <section
          class="search-section"
          id="overview"
        >
          <form
            class="search-form"
          >
            <div
              class="split-search"
            >
              <div
                class="left-half"
              >
                <div
                  class="searchbar-container"
                >
                  <label
                    for="command"
                  >
                    Enter Reason or Subject:
                  </label>
                  <input
                    autocomplete="off"
                    id="search-input"
                    name="command"
                    placeholder="Reason/Subject..."
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="searchbar-container"
                >
                  <label
                    for="owner"
                  >
                    Owners:
                  </label>
                  <input
                    autocomplete="off"
                    id="owner-input"
                    name="owner"
                    placeholder="Owner..."
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="or-divider"
              >
                <div
                  class="vertical-line"
                />
                <div
                  class="or"
                >
                  OR
                </div>
                <div
                  class="vertical-line"
                />
              </div>
              <div
                class="right-half"
              >
                <div
                  class="searchbar-container2"
                >
                  <label
                    for="team"
                  >
                    Team:
                  </label>
                  <input
                    autocomplete="off"
                    id="team-input"
                    name="team"
                    placeholder="Team..."
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <button
              id="toggle-advanced"
              type="button"
            >
              <b>
                Advanced Settings
              </b>
            </button>
            <button
              id="submitButton"
              type="submit"
            >
              <b>
                Search
              </b>
            </button>
          </form>
        </section>
        <section
          class="kpi-section"
          id="kpi-section"
        >
          <div
            class="kpi-card"
            data-email="team1@example.com"
          >
            <h3>
              Team: 
              Alpha
            </h3>
            <p>
              Total Commits: 
              <span>
                123
              </span>
            </p>
          </div>
          <div
            class="kpi-card"
            data-email="team2@example.com"
          >
            <h3>
              Team: 
              Beta
            </h3>
            <p>
              Open Reviews: 
              <span>
                8
              </span>
            </p>
          </div>
          <div
            class="kpi-card"
            data-email="team3@example.com"
          >
            <h3>
              Team: 
              Gamma
            </h3>
            <p>
              Bugs Fixed: 
              <span>
                15
              </span>
            </p>
          </div>
        </section>
        <section
          class="charts-section"
          id="charts-section"
        >
          <div
            class="chart-card"
          >
            <h3>
              Commit Trends
            </h3>
            <p>
              [Chart Placeholder]
            </p>
          </div>
          <div
            class="chart-card"
          >
            <h3>
              Code Quality Metrics
            </h3>
            <p>
              [Chart Placeholder]
            </p>
          </div>
        </section>
        <section
          class="table-section"
          id="table-section"
        >
          <div
            class="table-card"
          >
            <h3>
              Review Summary
            </h3>
            <p>
              [Table Placeholder]
            </p>
          </div>
        </section>
      </main>
    </div>
    <footer
      class="footer"
    >
      <p>
        Â© 2024 Team Metrics Dashboard @ elaneek - Henry Lee Ericsson
      </p>
    </footer>
  </div>
</div>
`;
